### ISSUE TITLE
vCard and xCard, --10 is October or 10 seconds?

### ISSUE THREAD

**Issue by Hywan** [fruux/sabre-vobject#198]
According to [RFC6351](http://tools.ietf.org/html/rfc6351#appendix-A), the `<date-and-or-time>` is of the form:
```
# 4.3.1
value-date = element date {
    xsd:string { pattern = "\d{8}|\d{4}-\d\d|--\d\d(\d\d)?|---\d\d" }
  }

# 4.3.2
value-time = element time {
    xsd:string { pattern = "(\d\d(\d\d(\d\d)?)?|-\d\d(\d\d?)|--\d\d)"
                         ~ "(Z|[+\-]\d\d(\d\d)?)?" }
  }

# 4.3.3
value-date-time = element date-time {
    xsd:string { pattern = "(\d{8}|--\d{4}|---\d\d)T\d\d(\d\d(\d\d)?)?"
                         ~ "(Z|[+\-]\d\d(\d\d)?)?" }
  }

# 4.3.4
value-date-and-or-time = value-date | value-date-time | value-time
```
While this format are a little bit crazy, I assume this is the format from ISO.8601.2004 according to [RFC6350](http://tools.ietf.org/html/rfc6350#section-4.3). Yes, there is no link on ISO.8601.2004 in my comment because ISO documents are not free :imp:! So this is my guess because these formats are very close based on different examples in RFC6350.

So what's the problem? Well, `--10` can fit into `value-date` **and** `value-time`:
  * From `value-date`, the 3rd element in the disjunction is `--\d\d(\d\d)?`, so it matches `--10`,
  * From `value-time`, the last element in the first disjunction is `--\d\d`, so it matches `--10`.

It's up to us to decide? This is a technical error in the RFC and an important one. We have therefore a bug in the `DateTimeParser` class. I will try to fix it. I assume the parsing of vCard date formats was empirical because the ISO.8601.2004 document is closed. Now, with the Relax NG Schema of xCard (RFC6351), we can something better.

**Comment by Hywan** [comment]
I have to check if jCard is impacted or not.

**Comment by Hywan** [comment]
Actually, when I say that `--10` can fit into `value-date` and `value-time`, it's not true, because `value-date-and-or-time` gives priority to `value-date`, so `--10` will always be October. But this is still an error in the RFC.

**Comment by evert** [comment]
How does the current date-and-or-time parser treat it ? :)

**Comment by Hywan** [comment]
@evert We have different functions for that. And it fails to parse 3 or 4 formats.

**Comment by Hywan** [comment]
Here is the new parser: https://github.com/Hywan/sabre-vobject/commit/50db424693fcd98dc2e4300e81f449e2e5fa684d. I will try to update the code to only rely on this method. It won't take a long time.

**Comment by Hywan** [comment]
I will update the code in another PR, after #192.

**Comment by Hywan** [comment]
[Erratum ID 4247](http://www.rfc-editor.org/errata_search.php?rfc=6351&eid=4247): there is an error in the `value-time` pattern (just a typo).

I didn't submit a report about this issue.

**Comment by Hywan** [comment]
[Erratum ID 4247](http://www.rfc-editor.org/errata_search.php?rfc=6351&eid=4247) has been verified.

The question remains for this issue: Should we submit an erratum or not?

**Comment by evert** [comment]
Report it =) Curious what the official response is.

**Comment by Hywan** [comment]
We can't really ask a question so here is my “erratum”: http://www.rfc-editor.org/errata_search.php?rfc=6350&eid=4261.

**Comment by evert** [comment]
Closing this. It looks like the erratum was responded to.
